### ISSUE TITLE
Using half-closed connections in Http().serverLayer

### ISSUE THREAD

**Issue by Quizzie** [akka/akka#19373]
In Akka HTTP version 2.0.1 the `Http().serverLayer` doesn't work correctly. If one sends an HTTP request through the `Http().serverLayer` and then an HTTP response in the other direction, the stream completes before the response is passed through.

A demonstration is shown in this project: https://github.com/Quizzie/server-layer-bug
The `bug.serverlayer.ServerLayerTest` is a test which uses the `Http().serverLayer` as an HTTP parser. 
The test in the `master` branch shows that it works correctly in Akka HTTP version 1.0.
The test in the `version-2.0.1` branch shows that it breaks in Akka HTTP version 2.0.1 (the test fails).

**Comment by ktoso** [comment]
Thanks for reporting and the reproducer, will have a look soon.
// cc @sirthias @jrudolph

**Comment by jrudolph** [comment]
@Quizzie you rely on a half-closed TCP connection (or stream in your case) to work with HTTP. This is currently not supported by akka-http.

I.e. it will not work to [close the input byte stream](https://github.com/Quizzie/server-layer-bug/blob/master/src/test/scala/bug/serverlayer/ServerLayerTest.scala#L94) before having received the response.

**Comment by MichaelZinsmaier** [comment]
After reading more about half-closed TCP and HTTP [related Akka Client Ticket](https://github.com/akka/akka/issues/16510), [2001 W3C post](https://lists.w3.org/Archives/Public/ietf-http-wg-old/2001JanApr/0037.html), ...
I agree, the HTTP spec seems to be surprisingly vague regarding half closed connections

=> maybe its not a bug its a **feature request** (-:


#### justification
tldr. a type A HTTP server is more defensive and standalone usage of the serverLayer would benefit
 
 
##### From a HTTP perspective

If there are major HTTP server implementations that treat half close as abort client implementations should not depend on it (they should be type 2 like the Akka HTTP client)

  * however following the arguments from the [2001 W3C post](https://lists.w3.org/Archives/Public/ietf-http-wg-old/2001JanApr/0037.html) a type A server would be more defensive (potentially with some performance cost)
  * Akka HTTP as integration layer might face some old/legacy/strange clients (in contrast to browsers / standard library based nice clients)

##### From a stream perspective:

The docs state that:

"... [the server Layer] can be useful ... to "run" the HTTP layer (and, potentially, higher-layers) against data that do not come from the network but rather some other source. Potential scenarios where this might be useful include tests, debugging or low-level event-sourcing (e.g by replaying network traffic)."

Ok lets assume I record HTTP requests into files and want to replay them later on. The problem is that there is no client application and thus no out of band knowledge about the number of  responses I expect.

* Scenario 1: I recorded one HTTP request
Reading it in with a FileSource and plugging that into the ServerLayer would not work out of the box because complete would have to be delayed until the response has been received.

* Scenario 2: I record two HTTP requests on a persisted connection
This is more difficult because I somehow have to know that I want to see two responses. But all the HTTP data (nr of requests, potential CLOSE headers, ...) is only available after parsing => inside the serverLayer

=> I suggest either implementing the serverLayer as type A server or having a more low level API that splits up parsing and HTTP protocol handling (to be able to inject some custom code)

aside from that having a type B server is proably just fine in most cases (-:

Best Michael

**Comment by jrudolph** [comment]
I don't disagree with your assessment. I guess one reason we disabled it at some point is that figuring out  closing conditions correctly is often quite hard and it's easy to create leaks this way.

**Comment by ktoso** [comment]
In any case – removing bug marker that that's cleared out.

**Comment by Quizzie** [comment]
Alright, let's consider this issue to be a feature request.
Should I close it or leave it open?

**Comment by ktoso** [comment]
FYI, half closed stuff again @johanandren - WDYT?

**Comment by johanandren** [comment]
Not sure I have an opinion, seems hard to get right, and this one in particular probably won't cause problems for many people. Maybe just make sure to document that this is how it works, unless someone wants to contribute a solution?

**Comment by jrudolph** [comment]
Ticket moved to http://github.com/akka/akka-http
 Rationale for the move discussed here: akka/akka-meta#27

If you are interested in contributing or tracking this issue, please comment in akka-http instead from now on :-)
